README.rst

===================
BaseX Client Driver
===================

A Python client driver for the BaseX_ XML database. **(Pre-Alpha! Not yet usable!)**

This driver communicates with a BaseX database running in `Server Mode`_ using
the native `BaseX Server Protocol`_ using a Pythonic API rather than the
default Java-flavored API of the `official python driver`_.

**Note** this software is not ready! (You can connect to a database, but there are
no session or query abstractions yet.) Maybe in a few weeks it will
be awesome, but in the meantime use the `official driver`_, which isn't pretty
but which works!


.. _BaseX: http://basex.org
.. _BaseX Server Protocol: http://docs.basex.org/wiki/Server_Protocol
.. _Server Mode: http://docs.basex.org/wiki/Startup#BaseX_Server
.. _official python driver: https://github.com/BaseXdb/basex-api/tree/master/src/main
.. _official driver: `official python driver`_


CHANGES.txt
v0.1, 2013-04-18 -- Initial release.


basexclient/test/socket_test.py
# -*- coding: utf-8 -*-
"""Test socket code"""
from .. import socket
from StringIO import StringIO
from nose.tools import eq_


def test_dataslices():
    # cases = [(holes, collection_length, expected), ...]
    cases = [
        ([1], 0, []),             # zero length
        ([], 10, [(0, 10)]),      # no holes
        ([], None, [(0, None)]),  # no holes, no length
        ([0], 11, [(1, 11)]),     # hole at start
        ([10], 11, [(0, 10)]),    # hole at end
        ([10], None, [(0, 10), (11, None)]),  # no length
        ([1], 11, [(0, 1), (2, 11)]),         # hole in middle
        # holes at beginning, middle, and end:
        ([0, 5, 8, 13], 14, [(1, 5), (6, 8), (9, 13)]),
        # consecutive holes:
        ([1, 2, 6, 10, 11, 12, 16], 17, [(0, 1), (3, 6), (7, 10), (13, 16)]),
        # consecutive and terminal holes:
        ([0, 1, 3, 4, 8, 12, 13, 14, 18, 19, 20], 21,
            [(2, 3), (5, 8), (9, 12), (15, 18)]),
    ]
    for holes, clen, expect in cases:
        yield check_dataslices, holes, clen, expect


def check_dataslices(holes, clen, expect):
    got = socket.dataslices(holes, clen)
    eq_(got, expect)


def test_dataslices_unsorted_holes():
    holes = [2, 1]
    try:
        socket.dataslices(holes)
    except ValueError:
        pass


def test_repack():
    # expected value for each string is a bytearray with underscores removed
    cases = [
        b'0123456789',
        b'_0123456789',
        b'0123456789_',
        b'0_123456789',
        b'0_123456789',
        b'0_123_456789',
        b'0_123_45_6789_',
        b'012345678_9_',
        b'__0123456789',
        b'0__123456789',
        b'__0__123_456___789__',
    ]
    for case in cases:
        ba = bytearray(case)
        holes = [i for i, c in enumerate(case) if c == b'_']
        expected_packed = case.replace(b'_', b'')
        expected_retval = len(expected_packed)
        yield check_repack, ba, holes, expected_packed, expected_retval


def check_repack(ba, holes, expected_packed, expected_retval):
    ba_copy = ba[:]
    got = socket.repack(ba_copy, holes)
    eq_(got, expected_retval)
    eq_(ba_copy[:expected_retval], expected_packed)


def test_unescape_bytearray():
    cases = [
        (b'1', b'1'),                # no escapes
        (b'\xFF', b'\xFF'),          # false positive
        (b'\xFF1234', b'\xFF1234'),  # false positive
        (b'\xFF\x001234', b'\x001234'),
        (b'12\xFF4\x00\xFF\x007', b'12\xFF4\x00\x007'),
    ]
    for arg, rv in cases:
        yield check_unescape_bytearray, bytearray(arg), rv


def check_unescape_bytearray(ba, expected):
    ba_copy = ba[:]
    packedlen = socket.unescape_bytearray(ba_copy)
    eq_(ba_copy[:packedlen], expected)


def test_escape_bytearray():
    cases = [
        (b'', b''),
        (b'12345', b'12345'),  # no escapes
        (b'\xFF', b'\xFF\xFF'),
        (b'\x001234', b'\xFF\x001234'),
        (b'\xFF\x001234', b'\xFF\xFF\xFF\x001234'),
        (b'12\xFF4\x00\xFF\x007', b'12\xFF\xFF4\xFF\x00\xFF\xFF\xFF\x007'),
    ]
    for arg, expected in cases:
        yield check_escape_bytearray, bytearray(arg), 0, None, expected


def check_escape_bytearray(ba, start, end, expected_escaped):
    ba_copy = ba[:]
    replacements = socket.escape_bytearray(ba_copy, start, end)
    eq_(replacements, len(ba_copy) - len(ba))
    eq_(ba_copy, expected_escaped)


def test_escape_bytearray_slice():
    # input, start, end, expected
    cases = [
        (b'', 1, 0, b''),  # impossible slicing
        (b'012\x0045\xFF', 3, None, b'012\xFF\x0045\xFF\xFF'),
        (b'\xFF12\x0045\xFF', 0, 4, b'\xFF\xFF12\xFF\x0045\xFF'),
        (b'\x0012\xFF\xFF45\x00', 3, 5, b'\x0012\xFF\xFF\xFF\xFF45\x00'),
    ]
    for arg, start, end, expected in cases:
        yield check_escape_bytearray, bytearray(arg), start, end, expected


def test_next_null():
    cases = [
        (b'12345', None),
        (b'\xFF\x00345', None),
        (b'12\xFF\x005', None),
        (b'123\xFF\x00', None),
        (b'\x002345', 1),
        (b'1\x00345', 2),
        (b'1234\x00', 5),
        (b'\xFF\x003\x005', 4),
        (b'\xFF\x0034\x00', 5),
        (b'1\x0034\x00', 2),
        (b'\x002\xFF\x005\x00', 1),
    ]
    for s, expect in cases:
        yield check_next_null, s, expect


def check_next_null(s, expected):
    found = socket.next_null(s)
    eq_(found, expected)


def test_bounded_buffer():
    s = b'0123456789'

    def readinto(buf):
        buf[:len(s)] = s
        return len(s)

    bb = socket.BoundedBuffer()
    assert isinstance(bb, bytearray)
    assert bb.isempty()
    bb.readintome(readinto)
    assert bb.datalen == 10
    assert not bb.isempty()
    assert isinstance(bb.view(), memoryview)
    assert bb.view() == s
    assert bb.view(5) == s[:5]
    assert bb.view(100) == s
    bb.slide_to(4)
    assert bb.datalen == 6
    assert bb.view() == s[4:]
    assert bb.view(100) == s[4:]
    assert not bb.isempty()
    bb.readintome(readinto)
    assert bb.view() == s
    bb.reset()
    assert bb.view() == b''
    assert bb.isempty()


class MockSocket(object):
    def __init__(self, data):
        self._data = StringIO(data)
        self._closed = False

    def recv_into(self, buf):
        maxbytes = len(buf)
        chunk = self._data.read(maxbytes)
        buf[:len(chunk)] = chunk
        return len(chunk)

    def close(self):
        self._closed = True


basexclient/test/basexsocket_test.py
# -*- coding: utf-8 -*-
"""Test basexsocket code"""
from .. import basexsocket
from StringIO import StringIO
from nose.tools import eq_, ok_, raises
from itertools import chain


def test_dataslices():
    # cases = [(holes, collection_length, expected), ...]
    cases = [
        ([1], 0, []),             # zero length
        ([], 10, [(0, 10)]),      # no holes
        ([], None, [(0, None)]),  # no holes, no length
        ([0], 11, [(1, 11)]),     # hole at start
        ([10], 11, [(0, 10)]),    # hole at end
        ([10], None, [(0, 10), (11, None)]),  # no length
        ([1], 11, [(0, 1), (2, 11)]),         # hole in middle
        # holes at beginning, middle, and end:
        ([0, 5, 8, 13], 14, [(1, 5), (6, 8), (9, 13)]),
        # consecutive holes:
        ([1, 2, 6, 10, 11, 12, 16], 17, [(0, 1), (3, 6), (7, 10), (13, 16)]),
        # consecutive and terminal holes:
        ([0, 1, 3, 4, 8, 12, 13, 14, 18, 19, 20], 21,
            [(2, 3), (5, 8), (9, 12), (15, 18)]),
    ]
    for holes, clen, expect in cases:
        yield check_dataslices, holes, clen, expect


def check_dataslices(holes, clen, expect):
    got = basexsocket.dataslices(holes, clen)
    eq_(got, expect)


@raises(ValueError)
def test_dataslices_unsorted_holes():
    holes = [2, 1]
    basexsocket.dataslices(holes)


def test_repack():
    # expected value for each string is a bytearray with underscores removed
    cases = [
        b'0123456789',
        b'_0123456789',
        b'0123456789_',
        b'0_123456789',
        b'0_123456789',
        b'0_123_456789',
        b'0_123_45_6789_',
        b'012345678_9_',
        b'__0123456789',
        b'0__123456789',
        b'__0__123_456___789__',
    ]
    for case in cases:
        ba = bytearray(case)
        holes = [i for i, c in enumerate(case) if c == b'_']
        expected_packed = case.replace(b'_', b'')
        expected_retval = len(expected_packed)
        yield check_repack, ba, holes, expected_packed, expected_retval


def check_repack(ba, holes, expected_packed, expected_retval):
    ba_copy = ba[:]
    got = basexsocket.repack(ba_copy, holes)
    eq_(got, expected_retval)
    eq_(ba_copy[:expected_retval], expected_packed)


def test_unescape_bytearray():
    cases = [
        (b'1', b'1'),                # no escapes
        (b'\xFF', b'\xFF'),          # false positive
        (b'\xFF1234', b'\xFF1234'),  # false positive
        (b'\xFF\x001234', b'\x001234'),
        (b'12\xFF4\x00\xFF\x007', b'12\xFF4\x00\x007'),
    ]
    for arg, rv in cases:
        yield check_unescape_bytearray, bytearray(arg), rv


def check_unescape_bytearray(ba, expected):
    ba_copy = ba[:]
    packedlen = basexsocket.unescape_bytearray(ba_copy)
    eq_(ba_copy[:packedlen], expected)


def test_escape_bytearray():
    cases = [
        (b'', b''),
        (b'12345', b'12345'),  # no escapes
        (b'\xFF', b'\xFF\xFF'),
        (b'\x001234', b'\xFF\x001234'),
        (b'\xFF\x001234', b'\xFF\xFF\xFF\x001234'),
        (b'12\xFF4\x00\xFF\x007', b'12\xFF\xFF4\xFF\x00\xFF\xFF\xFF\x007'),
    ]
    for arg, expected in cases:
        yield check_escape_bytearray, bytearray(arg), 0, None, expected


def check_escape_bytearray(ba, start, end, expected_escaped):
    ba_copy = ba[:]
    replacements = basexsocket.escape_bytearray(ba_copy, start, end)
    eq_(replacements, len(ba_copy) - len(ba))
    eq_(ba_copy, expected_escaped)


def test_escape_bytearray_slice():
    # input, start, end, expected
    cases = [
        (b'', 1, 0, b''),  # impossible slicing
        (b'012\x0045\xFF', 3, None, b'012\xFF\x0045\xFF\xFF'),
        (b'\xFF12\x0045\xFF', 0, 4, b'\xFF\xFF12\xFF\x0045\xFF'),
        (b'\x0012\xFF\xFF45\x00', 3, 5, b'\x0012\xFF\xFF\xFF\xFF45\x00'),
    ]
    for arg, start, end, expected in cases:
        yield check_escape_bytearray, bytearray(arg), start, end, expected


def test_next_null():
    cases = [
        (b'12345', None),
        (b'\xFF\x00345', None),
        (b'12\xFF\x005', None),
        (b'123\xFF\x00', None),
        (b'\x002345', 1),
        (b'1\x00345', 2),
        (b'1234\x00', 5),
        (b'\xFF\x003\x005', 4),
        (b'\xFF\x0034\x00', 5),
        (b'1\x0034\x00', 2),
        (b'\x002\xFF\x005\x00', 1),
    ]
    for s, expect in cases:
        yield check_next_null, s, expect


def check_next_null(s, expected):
    found = basexsocket.next_null(s)
    eq_(found, expected)


def test_bounded_buffer():
    s = b'0123456789'

    def readinto(buf):
        buf[:len(s)] = s
        return len(s)

    bb = basexsocket.BoundedBuffer()
    assert isinstance(bb, bytearray)
    assert bb.isempty()
    bb.readintome(readinto)
    eq_(bb.datalen, 10)
    assert not bb.isempty()
    assert isinstance(bb.view(), (buffer, memoryview))
    eq_(bb.view(),  s)
    eq_(bb.view(5), s[:5])
    assert bb.view(100) == s
    bb.slide_to(4)
    eq_(bb.datalen,  6)
    eq_(bb.view(),  s[4:])
    eq_(bb.view(100),  s[4:])
    eq_(bb.view(1),  s[4])
    assert not bb.isempty()
    bb.readintome(readinto)
    eq_(bb.view(), s)
    bb.reset()
    eq_(bb.view(), b'')
    assert bb.isempty()


class MockSocket(object):
    def __init__(self, data):
        self._data = StringIO(data)
        self._closed = False

    def recv_into(self, buf):
        maxbytes = len(buf)
        chunk = self._data.read(maxbytes)
        buf[:len(chunk)] = chunk
        return len(chunk)

    def close(self):
        self._closed = True


def create_bufferedsocket(data):
    soc = lambda: MockSocket(data)
    return basexsocket.BufferedSocket(soc)


def test_bufferedsocket_read_next():
    data = b'0123456789\x00\x01'
    with create_bufferedsocket(data) as bs:
        eq_(bs.read_byte(), b'0')
        eq_(bs.read_next(), b'123456789\x00')
        eq_(bs.read_byte(), b'\x01')


def test_bufferedsocket_multiboundary():
    """Test how bufferedsocket works when multiple socket reads are needed"""
    trailer =  b'\x0012\xFF\xFF345\x00\x01'
    datagens = chain(
        (b'_' for i in xrange(0xfff)),
        b'X',
        (b'_' for i in xrange(0xfff)),
        b'X',
        (b'_' for i in xrange(0xfff)),
        b'X',
        (b'_' for i in xrange(0xfff)),
        b'X',
        trailer
    )
    data = bytearray(datagens)
    with create_bufferedsocket(data) as bs:
        eq_(bs.read_next(), data[:0x4001])
        eq_(bs.read_next(), trailer[1:9])

def test_bufferedsocket_read_iter():
    data = b'01234\xFF\x00\x0056789\x00\x01'
    with create_bufferedsocket(data) as bs:
        eq_(list(bs.read_next_iter()), [b'01234\xFF\x00\x00'])

    data = bytearray(chain((b'X' for i in xrange(0x4000)), b'\x00\x00'))
    with create_bufferedsocket(data) as bs:
        chunkiter = bs.read_next_iter()
        chunk = chunkiter.next()
        ok_(len(chunk) < 0x4000)
        ok_(all(c == b'X' for c in chunk))
        # round 2
        chunk = chunkiter.next()
        ok_(len(chunk) < 0x4000)
        ok_(all(c == b'X' for c in chunk))
        # last chunk
        for c in chunkiter:
            chunk = c
        ok_(len(chunk) < 0x4000)
        ok_(all(c == b'X' for c in chunk[:-1]))
        ok_(c[-1] == b'\x00')







basexclient/basexsocket.py
# -*- coding: utf-8 -*-
"""Lowest-level BaseX client-server communication"""

import re
import io
import socket

C00 = b'\x00'
CFF = b'\xFF'
I00 = 0
IFF = 255

r_escapeable = re.compile(b'[{}{}]'.format(C00, CFF))
r_escaped = re.compile(b'{}[{}{}]'.format(CFF, C00, CFF))
r_unescaped_null = re.compile(b'(?<!{}){}'.format(CFF, C00))

class Commands(object):
    QUERY = b'\x00'
    CLOSE = b'\x02'
    BIND = b'\x03'
    RESULTS = b'\x04'
    EXECUTE = b'\x05'
    INFO = b'\x06'
    OPTIONS = b'\x07'
    CREATE = b'\x08'
    ADD = b'\x09'
    WATCH = b'\x0A'
    UNWATCH = b'\x0B'
    REPLACE = b'\x0C'
    STORE = b'\x0D'
    CONTEXT = b'\x0E'
    UPDATING = b'\x1E'
    FULL = b'\x1F'


class BoundedBuffer(bytearray):
    def __init__(self, size=io.DEFAULT_BUFFER_SIZE):
        super(BoundedBuffer, self).__init__(size)
        self.dataslice = slice(0, 0)

    def reset(self):
        self.dataslice = slice(0, 0)

    @property
    def datalen(self):
        ds = self.dataslice
        return ds.stop-ds.start

    def slide_to(self, idx):
        if idx is None:
            self.reset()
        else:
            self.dataslice = slice(self.dataslice.start + idx, self.dataslice.stop)

    def view(self, nbytes=None):
        """Return memoryview of next bytes without consuming them.

        ``nbytes`` is the max number of bytes to return (maybe fewer).
        """
        if nbytes is None:
            nbytes = len(self)
        endidx = min(self.dataslice.stop, self.dataslice.start + nbytes)
        return memoryview(self)[self.dataslice.start:endidx]

    def view_buf(self, nbytes=None):
        """Return buffer of next bytes without consuming them.

        Only needed by re objects in Python 2.7, which can't search
        a memoryview
        """
        if nbytes is None:
            nbytes = len(self)
        maxlen = min(self.dataslice.stop-self.dataslice.start, nbytes)
        return buffer(self, self.dataslice.start, maxlen)

    def isempty(self):
        return not self.datalen

    def readintome(self, readintomethod):
        bytesread = readintomethod(self)
        #pylint: disable=W0201
        self.dataslice = slice(0, bytesread)


class BufferedSocket(object):
    """Buffers a socket

    This is *not* a threadsafe object--do not share!
    """
    def __init__(self, socketfactory):
        """Initialize the socket wrapper

        socketfactory -- a callable which returns a socket object
                         which this object will wrap
        """
        self._soc = socketfactory()
        self._buf = BoundedBuffer()

    @classmethod
    def connect(cls, address):
        """Connect to address (host, port)"""
        return cls(lambda: socket.create_connection(address))

    def close(self):
        return self._soc.close()

    def __enter__(self):
        return self

    def __exit__(self, type_, value, traceback):
        self.close()
        return False

    def _read_next_iter(self):
        """Yield memoryviews of buffers up to and including a null terminator.

        Used to implement recv_next_iter() and recv_next(). Use carefully:
        returned memoryviews must be materialized before the next yield!
        """
        do_readinto = self._soc.recv_into
        buf = self._buf
        while True:
            if buf.isempty():
                buf.readintome(do_readinto)
            found = next_null(buf.view_buf())
            assert found != 0
            # if found is None, whole buffer is copied
            yield buf.view(found)
            buf.slide_to(found)
            if found is not None:
                return

    def read_next_iter(self):
        """Yield bytes from socket up to and including a null terminator.

        This method is the same as recv_next() except it yields bytes as they
        are available instead of buffering them all into a bytearray. This
        can reduce latency for large responses, but it will be less efficient.
        """
        for mv in self._read_next_iter():
            yield mv.tobytes()

    def read_next(self):
        """Receive bytes up to and including an unescaped null terminator"""
        # This looks strange, but it's to reduce fragmentation
        # Almost always we will only have to grab one chunk, so we allocate
        # its memory all at once instead of growing an empty bytearray.
        ba = None
        for chunk in self._read_next_iter():
            if ba is None:
                ba = bytearray(chunk)
            else:
                ba.extend(chunk)
        return ba if ba is not None else bytearray()

    def read_byte(self):
        """Return the next byte in the socket"""
        do_readinto = self._soc.recv_into
        buf = self._buf
        if buf.isempty():
            buf.readintome(do_readinto)
        nextbyte = buf.view(1).tobytes()
        buf.slide_to(1)
        return nextbyte

    def write(self, data):
        """Return bytes of data sent in one system call."""
        return self._soc.send(data)

    def write_all(self, data):
        """Write all data."""
        self._soc.sendall(data)


def next_null(ba, r_null=r_unescaped_null.search):
    """Return memoryview of bytes up to and including the next unescaped null

    Returns None if not found."""
    m = r_null(ba)
    if not m:
        return None
    return m.end()


def escape_bytearray(ba, start=0, end=None, r_escapeable=r_escapeable, IFF=IFF):
    """Return number of escaped bytes after in-place escape of characters

    If start or end are supplied, will escape only data in that slice. Any
    bytes before or after will be untouched, but the index of following bytes
    may increase.

    Return value will be number of bytes added.

    """
    #pylint: disable=W0621
    if end is None:
        end = len(ba)
    findescapeable = lambda s, e, ba=ba, srch=r_escapeable.search: srch(ba, s, e)
    n_matches = 0
    while True:
        m = findescapeable(start, end)
        if m is None:
            break
        ridx, start = m.span()
        ba.insert(ridx, IFF)
        start += 1
        end += 1
        n_matches += 1
    return n_matches


def unescape_bytearray(ba, r_escaped=r_escaped):
    """Return length of data after in-place removal of escape characters"""
    #pylint: disable=W0621
    # need at least two characters for an escape sequence
    if len(ba) < 2:
        return len(ba)
    toremove = [m.start() for m in r_escaped.finditer(ba)]
    return repack(ba, toremove)


def repack(ba, holes):
    """Repack bytes in a bytearray removing indexes in sorted list "holes"

    Returns the slice offset where the packed data ends, which is always
    len(ba)-len(holes).
    """
    lenba = len(ba)
    if not holes:
        return lenba
    mv = memoryview(ba)
    dslices = dataslices(holes, lenba)
    pstart = 0
    for dstart, dend in dslices:
        pend = pstart + dend - dstart
        mv[pstart:pend] = mv[dstart:dend]
        pstart = pend
    return pend


def dataslices(holes, clen=None):
    """Return a list of slices between sorted list of hole indexes

    holes must be pre-sorted!

    clen is the length of the collection. If not provided a slice of (_, None)
    will always be appended.
    """
    #pylint: disable=C0301
    if clen <= 0 and clen is not None:
        return []
    if not holes:
        return [(0, clen)]
    orig_holes = holes
    dslices = []
    if holes[0] > 0:
        dslices.append((0, holes[0]))
        lasthole = holes[0] + 1
        holes = holes[1:]
    else:
        lasthole = 1
    for hole in holes:
        if hole > lasthole:
            dslices.append((lasthole, hole))
        elif hole < lasthole - 1:
            raise ValueError('holes argument is not sorted: {!r}'.format(orig_holes))
        lasthole = hole + 1
    if lasthole < clen or clen is None:
        dslices.append((lasthole, clen))
    return dslices


basexclient/socket.py
# -*- coding: utf-8 -*-
"""Lowest-level BaseX client-server communication"""

import re
import io
import socket

C00 = b'\x00'
CFF = b'\xFF'
I00 = 0
IFF = 255

r_escapeable = re.compile(b'[{}{}]'.format(C00, CFF))
r_escaped = re.compile(b'{}[{}{}]'.format(CFF, C00, CFF))
r_unescaped_null = re.compile(b'(?<!{}){}'.format(CFF, C00))


class Commands(object):
    QUERY = b'\0'
    CLOSE = b'\2'
    BIND = b'\3'
    RESULTS = b'\4'
    EXECUTE = b'\5'
    INFO = b'\6'
    OPTIONS = b'\7'
    CREATE = b'\8'
    ADD = b'\9'
    WATCH = b'\10'
    UNWATCH = b'\11'
    REPLACE = b'\12'
    STORE = b'\13'
    CONTEXT = b'\14'
    UPDATING = b'\30'
    FULL = b'\31'


class BoundedBuffer(bytearray):
    def __new__(cls, size=io.DEFAULT_BUFFER_SIZE):
        return bytearray.__new__(cls, size)

    def __init__(self):
        bytearray.__init__(self)
        self.reset()

    def reset(self):
        self.dataslice = slice(0, 0)

    @property
    def datalen(self):
        ds = self.dataslice
        return ds.stop-ds.start

    def slide_to(self, idx):
        self.dataslice = slice(self.dataslice.start + idx, self.dataslice.stop)

    def view(self, endat=None):
        if endat is None:
            endat = len(self)
        endidx = min(self.dataslice.stop, endat)
        return memoryview(self)[self.dataslice.start:endidx]

    def isempty(self):
        return not self.datalen

    def readintome(self, readintomethod):
        bytesread = readintomethod(self)
        self.dataslice = slice(0, bytesread)


class BufferedSocket(object):
    """Buffers a socket

    This is *not* a threadsafe object--do not share!
    """
    def __init__(self, socketfactory):
        """Initialize the socket wrapper

        socketfactory -- a callable which returns a socket object
                         which this object will wrap
        """
        self._soc = socketfactory()
        self._buf = BoundedBuffer()

    @classmethod
    def connect(cls, address):
        """Connect to address (host, port)"""
        return cls(lambda: socket.create_connection(address))

    def close(self):
        return self._soc.close()

    def __enter__(self):
        return self

    def __exit__(self, type_, value, traceback):
        self.close()
        return False

    def _recv_next_iter(self):
        """Yield memoryviews of buffers up to and including a null terminator.

        Used to implement recv_next_iter() and recv_next(). Use carefully:
        returned memoryviews must be materialized before the next yield!
        """
        do_readinto = self._soc.recv_into
        buf = self._buf
        while True:
            if buf.isempty():
                buf.readintome(do_readinto)
            found = next_null(buf.view())
            assert found != 0
            # if found is None, whole buffer is copied
            yield buf.view(found)
            if found:
                buf.slide_to(found)
                return

    def recv_next_iter(self):
        """Yield bytes from socket up to and including a null terminator.

        This method is the same as recv_next() except it yields bytes as they
        are available instead of buffering them all into a bytearray. This
        can reduce latency for large responses, but it will be less efficient.
        """
        for mv in self._recv_next_iter():
            yield mv.tobytes()

    def recv_next(self):
        """Receive bytes up to and including an unescaped null terminator"""
        return bytearray(self._recv_next_iter())


def next_null(ba, r_null=r_unescaped_null.search):
    """Return a memoryview of bytes up to and including the next unescaped \\0 or None if not found"""
    m = r_null(ba)
    if not m:
        return None
    return m.end()


def escape_bytearray(ba, start=0, end=None, r_escapeable=r_escapeable, IFF=IFF):
    """Prefix escapeable characters in bytearray in-place; return number of escapes done

    If start or end are supplied, will escape only data in that slice. Any
    bytes before or after will be untouched, but the index of following bytes
    may increase.

    Return value will be number of bytes added.

    """
    if end is None:
        end = len(ba)
    findescapeable = lambda s, e, ba=ba, srch=r_escapeable.search: srch(ba, s, e)
    n_matches = 0
    while True:
        m = findescapeable(start, end)
        if m is None:
            break
        ridx, start = m.span()
        ba.insert(ridx, IFF)
        start += 1
        end += 1
        n_matches += 1
    return n_matches


def unescape_bytearray(ba, r_escaped=r_escaped):
    """Remove the escape characters from a bytearray in-place; return length of valid data"""
    # need at least two characters for an escape sequence
    if len(ba) < 2:
        return len(ba)
    toremove = [m.start() for m in r_escaped.finditer(ba)]
    return repack(ba, toremove)


def repack(ba, holes):
    """Repack bytes in a bytearray removing indexes in sorted list "holes"

    Returns the slice offset where the packed data ends, which is always
    len(ba)-len(holes).
    """
    lenba = len(ba)
    if not holes:
        return lenba
    mv = memoryview(ba)
    dslices = dataslices(holes, lenba)
    pstart = 0
    for i, (dstart, dend) in enumerate(dslices):
        pend = pstart + dend - dstart
        mv[pstart:pend] = mv[dstart:dend]
        pstart = pend
    return pend


def dataslices(holes, clen=None):
    """Return a list of slices between sorted list of hole indexes

    holes must be pre-sorted!

    clen is the length of the collection. If not provided a slice of (_, None)
    will always be appended.
    """
    if clen <= 0 and clen is not None:
        return []
    if not holes:
        return [(0, clen)]
    orig_holes = holes
    dslices = []
    if holes[0] > 0:
        dslices.append((0, holes[0]))
        lasthole = holes[0] + 1
        holes = holes[1:]
    else:
        lasthole = 1
    for hole in holes:
        if hole > lasthole:
            dslices.append((lasthole, hole))
        elif hole < lasthole - 1:
            raise ValueError('holes argument is not sorted: {!r}'.format(orig_holes))
        lasthole = hole + 1
    if lasthole < clen or clen is None:
        dslices.append((lasthole, clen))
    return dslices
